File: mseed/c/test/test_steim.c
/* SPDX-License-Identifier: MIT
 * Steim-1/2 codec tests: internal round-trip, byte-exact vs the Python reference,
 * overflow handling, and the constant-signal compression claim. */
#include "nic_mseed.h"
#include "vectors.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int passed = 0, failed = 0;
static void check(const char *name, int cond) {
    if (cond) { passed++; printf("  PASS  %s\n", name); }
    else      { failed++; printf("  FAIL  %s\n", name); }
}

#define MAXR 4096
#define MAXN 4096

/* encode the whole series record by record, decode each back, compare */
static int roundtrip(const int32_t *s, size_t n, int version, int fpr) {
    uint8_t rec[MAXR];
    int32_t dec[MAXN];
    size_t i = 0; int32_t prev = 0;
    while (i < n) {
        size_t used = 0;
        if (nic_steim_encode_record(s + i, n - i, version, fpr, prev, rec, &used) != 0) return 0;
        if (used == 0) return 0;
        if (nic_steim_decode_record(rec, (size_t)fpr * 64, used, version, dec) != 0) return 0;
        for (size_t k = 0; k < used; k++) if (dec[k] != s[i + k]) return 0;
        prev = s[i + used - 1];
        i += used;
    }
    return 1;
}

int main(void) {
    printf("Steim-1/2 codec tests (C)\n");

    /* ---- internal round-trip on deterministic series ---- */
    static int32_t buf[MAXN];

    int32_t single[1] = { 12345 };
    check("S1 round-trip: single", roundtrip(single, 1, NIC_STEIM1, 7));
    check("S2 round-trip: single", roundtrip(single, 1, NIC_STEIM2, 7));

    for (int i = 0; i < 50; i++) buf[i] = 777;
    check("S1 round-trip: constant", roundtrip(buf, 50, NIC_STEIM1, 7));
    check("S2 round-trip: constant", roundtrip(buf, 50, NIC_STEIM2, 7));

    for (int i = 0; i < 300; i++) buf[i] = i;
    check("S1 round-trip: ramp", roundtrip(buf, 300, NIC_STEIM1, 7));
    check("S2 round-trip: ramp", roundtrip(buf, 300, NIC_STEIM2, 7));

    /* pseudo-random walk (deterministic LCG), small then medium diffs */
    uint32_t st = 1u; int32_t acc = 0;
    for (int i = 0; i < 500; i++) { st = st * 1103515245u + 12345u;
        acc += (int32_t)((st >> 16) % 15) - 7; buf[i] = acc; }
    check("S1 round-trip: tiny diffs", roundtrip(buf, 500, NIC_STEIM1, 7));
    check("S2 round-trip: tiny diffs", roundtrip(buf, 500, NIC_STEIM2, 7));

    st = 99u; acc = 0;
    for (int i = 0; i < 500; i++) { st = st * 1103515245u + 12345u;
        acc += (int32_t)((st >> 12) % 6001) - 3000; buf[i] = acc; }
    check("S1 round-trip: medium diffs", roundtrip(buf, 500, NIC_STEIM1, 7));
    check("S2 round-trip: medium diffs", roundtrip(buf, 500, NIC_STEIM2, 7));

    st = 7u; acc = 0;
    for (int i = 0; i < 300; i++) { st = st * 1103515245u + 12345u;
        acc += (int32_t)((st >> 8) % 20000001) - 10000000; buf[i] = acc; }
    check("S1 round-trip: large 30-bit diffs", roundtrip(buf, 300, NIC_STEIM1, 7));
    check("S2 round-trip: large 30-bit diffs", roundtrip(buf, 300, NIC_STEIM2, 7));

    for (int i = 0; i < 400; i++) buf[i] = (i & 1 ? -1 : 1) * (i % 9);
    check("S1 round-trip: negatives + zeros", roundtrip(buf, 400, NIC_STEIM1, 7));
    check("S2 round-trip: negatives + zeros", roundtrip(buf, 400, NIC_STEIM2, 7));

    for (int i = 0; i < 200; i++) buf[i] = (i & 1) ? (1 << 20) : 1;
    check("S1 round-trip: alternating big/small", roundtrip(buf, 200, NIC_STEIM1, 7));
    check("S2 round-trip: alternating big/small", roundtrip(buf, 200, NIC_STEIM2, 7));

    /* ---- byte-exact vs the Python reference (ObsPy-validated) ---- */
    uint8_t rec[MAXR]; size_t used = 0;
    int rc2 = nic_steim_encode_record(VEC_A, sizeof(VEC_A) / sizeof(VEC_A[0]),
                                      NIC_STEIM2, 7, 0, rec, &used);
    check("S2 byte-exact vs Python ref (record)",
          rc2 == 0 && used == sizeof(VEC_A) / sizeof(VEC_A[0]) &&
          memcmp(rec, REC_S2, sizeof(REC_S2)) == 0);
    int rc1 = nic_steim_encode_record(VEC_A, sizeof(VEC_A) / sizeof(VEC_A[0]),
                                      NIC_STEIM1, 7, 0, rec, &used);
    check("S1 byte-exact vs Python ref (record)",
          rc1 == 0 && used == sizeof(VEC_A) / sizeof(VEC_A[0]) &&
          memcmp(rec, REC_S1, sizeof(REC_S1)) == 0);

    /* ---- overflow handling ---- */
    int32_t big[2] = { 0, (1 << 29) };           /* diff = 2^29 > 30-bit signed max */
    rc2 = nic_steim_encode_record(big, 2, NIC_STEIM2, 7, 0, rec, &used);
    check("S2 rejects >30-bit diff", rc2 == NIC_MSEED_EOVERFLOW);
    check("S1 accepts up to 32-bit diff", roundtrip(big, 2, NIC_STEIM1, 7));

    /* ---- constant signal compresses hard ---- */
    for (int i = 0; i < 1000; i++) buf[i] = 100;
    size_t i = 0, total = 0; int32_t prev = 0;
    while (i < 1000) {
        if (nic_steim_encode_record(buf + i, 1000 - i, NIC_STEIM2, 7, prev, rec, &used) != 0) break;
        total += 7 * 64; prev = buf[i + used - 1]; i += used;
    }
    check("constant signal: >=2x compression vs raw int32", total < 1000 * 4 / 2);
    check("constant signal round-trips", roundtrip(buf, 1000, NIC_STEIM2, 7));

    printf("\nResult: %d/%d PASS | %d FAIL\n", passed, passed + failed, failed);
    return failed ? 1 : 0;
}

File: mseed/tests/test_steim.py
# SPDX-License-Identifier: MIT
"""Round-trip + edge tests for the Steim-1/2 codec (no external deps)."""
import os
import random
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from nic_mseed.steim import STEIM1, STEIM2, encode, decode, encode_record, decode_record

_p = _f = 0
def check(name, cond):
    global _p, _f
    if cond: _p += 1; print(f"  PASS  {name}")
    else:    _f += 1; print(f"  FAIL  {name}")


def _roundtrip(samples, version, fpr=7):
    recs = encode(samples, version, frames_per_record=fpr)
    # recover per-record counts by decoding greedily: we know total; re-derive by
    # decoding each record with the max it could hold, then trim — instead, encode
    # exposes counts via encode_record, so reproduce the split here:
    counts, prev, i = [], 0, 0
    s = list(samples)
    for rec in recs:
        _, used = encode_record(s[i:], version, fpr, prev)
        counts.append(used); prev = s[i + used - 1]; i += used
    return decode(recs, counts, version)


def main():
    print("Steim-1/2 codec tests")
    random.seed(1)
    cases = {
        "single sample":        [12345],
        "constant":             [777] * 50,
        "small ramp":           list(range(0, 300)),
        "tiny diffs (4-bit)":   [0] + list(__import__("itertools").accumulate(
                                    [random.randint(-7, 7) for _ in range(500)])),
        "medium diffs":         [0] + list(__import__("itertools").accumulate(
                                    [random.randint(-3000, 3000) for _ in range(500)])),
        "large (15/30-bit)":    [0] + list(__import__("itertools").accumulate(
                                    [random.randint(-10_000_000, 10_000_000) for _ in range(300)])),
        "negatives + zeros":    [(-1) ** i * (i % 9) for i in range(400)],
        "alternating big/small":[(1 << 20) if i % 2 else 1 for i in range(200)],
    }
    for name, samples in cases.items():
        for ver, vn in ((STEIM1, "S1"), (STEIM2, "S2")):
            try:
                got = _roundtrip(samples, ver)
                check(f"{vn} round-trip: {name}", got == list(samples))
            except OverflowError as e:
                check(f"{vn} round-trip: {name} (overflow handled)", ver == STEIM2 and False or True) \
                    if False else check(f"{vn} {name}: unexpected overflow", False)

    # Steim-2 must reject a >30-bit difference; Steim-1 must accept up to 32-bit.
    big = [0, (1 << 29)]            # diff = 2^29 — fits 30-bit signed? max is 2^29-1 → no
    try:
        encode(big, STEIM2); check("S2 rejects >30-bit diff", False)
    except OverflowError:
        check("S2 rejects >30-bit diff", True)
    check("S1 accepts 32-bit diff", _roundtrip([0, (1 << 29)], STEIM1) == [0, (1 << 29)])

    # Frame math: a constant signal (all-zero diffs) packs 7×4-bit and compresses
    # hard — a 7-frame record holds 103 data words × 7 = 721 such samples.
    flat = [100] * 1000
    recs = encode(flat, STEIM2, frames_per_record=7)
    total = len(recs) * 7 * 64
    check("constant signal: ≥2× compression vs raw int32", total < len(flat) * 4 // 2)
    check("constant signal round-trips", _roundtrip(flat, STEIM2) == flat)

    print(f"\nResult: {_p}/{_p+_f} PASS | {_f} FAIL")
    sys.exit(0 if _f == 0 else 1)


if __name__ == "__main__":
    main()
